src/dune_engine/fs_memo.ml

```diff
@@ -155,11 +190,12 @@ let depend path =
 
    Currently, we do not expose this low-level primitive. If you need it, perhaps
    you could add a higher-level primitive instead, such as [path_exists]? *)
-let declaring_dependency path ~f =
-  let+ () = depend path in
+let declaring_dependency (kind, path) ~f =
+  let+ () = depend (kind, path) in
   f path
 
-let path_stat = declaring_dependency ~f:Fs_cache.(read Untracked.path_stat)
+let path_stat path =
```

Comment:
I'm not sure this is right. What if `path` turns out to be a directory? Then I think we need to subscribe to it directly rather than via the parent. 

Also, what if it's the root? Then there is no parent and this will just crash.

Reply:
(Same for `path_digest` I guess, since for now we are allowing to call it on directories.)

Reply:
> What if path turns out to be a directory? Then I think we need to subscribe to it directly rather than via the parent.

I think watching the parent for all stat changes is right. Though I could be wrong about that. It is only when we care for the contents of a directory that we watch the path directly

> Also, what if it's the root? Then there is no parent and this will just crash.

Hmm, I thought a pushed a fix for this. In any case, the correct behavior is just to watch root in this case.

Reply:
> I think watching the parent for all stat changes is right. 

How about the `st_mtime` field, for example? For directories, it's changed if its listing changes, but I doubt this information will be propagated all the way to the directory's parent. We do not watch `st_mtime` right now, but this might change. Perhaps, with the current `Reduced_stats`, watching directories via parent is OK, but I'm not very confident about this given the `st_mtime` example. How about `st_size` and `st_nlink`? If we include them into `Reduced_stats` one day, watching might get subtly broken. I think we should just always watch directories directly to avoid such pitfalls.

> Hmm, I thought a pushed a fix for this. In any case, the correct behavior is just to watch root in this case.

Agreed.

Reply:
Yeah, you've convinced me this is safer. Always watching directories (in addition to their parents) is the right way to do.